velox/exec/PartitionedOutputBufferManager.h

```diff
@@ -40,22 +40,65 @@ struct DataAvailable {
   }
 };
 
-class DestinationBuffer {
+/// The class is used to buffer the output pages which haven't been fetched by
+/// any destination for arbitrary output.
+///
+/// NOTE: there is only one arbitrary buffer setup for arbitrary output to share
+/// among destinations. Also, this class is not thread-safe.
+class ArbitraryBuffer {
  public:
-  void enqueue(std::shared_ptr<SerializedPage> data) {
-    // drop duplicate end markers
-    if (data == nullptr && !data_.empty() && data_.back() == nullptr) {
-      return;
-    }
+  /// Returns true if this arbitrary buffer has no buffered pages.
+  bool empty() const {
+    return pages_.empty() || hasNoMoreData();
+  }
 
-    data_.push_back(std::move(data));
+  /// Returns true if this arbitrary buffer will not receive any new pages from
+  /// enqueue() but it can still has buffered pages waiting to dispatch to
+  /// destination on data fetch.
+  bool hasNoMoreData() const {
+    return !pages_.empty() && (pages_.back() == nullptr);
   }
 
+  /// Marks this arbitrary buffer will not receive any new incoming pages. It
```

Comment:
NIT: Marks this arbitrary buffer to not receive new incoming pages.

And also move the comment of "It appends a null page at the end of 'pages_' as end marker." to the implementation of `noMoreData()`?  I think we should avoid putting detailed implementation comment here

Reply:
I think it is good to put detailed comments for internal component interface and this is not for public APIs which might not be that useful.